src/utils/stacktrace.ts

```diff
@@ -0,0 +1,95 @@
+
+export type RawStackFrames = string;
+
+const getErrorTitle = (e: Error): string => `${e.name || "Error"}: ${e.message}`;
+const getErrorStackFrames = (e: Error): string | undefined => e.stack?.replace(getErrorTitle(e) + "\n", "");
+
+export function captureRawStackFrames(filterFunc?: (...args: unknown[]) => unknown): RawStackFrames {
+    const savedStackTraceLimit = Error.stackTraceLimit;
+    const targetObj = {} as { stack: RawStackFrames };
+
+    Error.stackTraceLimit = 64;
+    Error.captureStackTrace(targetObj, filterFunc || captureRawStackFrames);
+    Error.stackTraceLimit = savedStackTraceLimit;
+
+    const rawFramesPosition = targetObj.stack.indexOf("\n") + 1; // crop out error message
+
+    return targetObj.stack.slice(rawFramesPosition);
+}
+
+export function applyStackFrames(error: Error, frames: RawStackFrames): Error {
+    if (!error || !error.message) {
+        return error;
+    }
+
+    error.stack = getErrorTitle(error) + "\n" + frames;
+
+    return error;
+}
+
+export function filterExtraWdioFrames(error: Error): Error {
+    if (!error || !error.message || !error.stack) {
+        return error;
+    }
+
+    const frames = getErrorStackFrames(error)!;
+    const rawFramesArr = frames.split("\n");
+    const framesParsed = ErrorStackParser.parse(error);
+
+    if (rawFramesArr.length !== framesParsed.length) {
+        return error;
+    }
+
+    const ignoredFunctions = [
+        "Browser.wrapCommandFn",
+        "Element.wrapCommandFn",
+        "Element.<anonymous>",
+        "Element.newCommand",
+        "Element.elementErrorHandlerCallbackFn",
+    ];
+
+    const isWdioFrame = (frame: StackFrame): boolean =>
+        Boolean(frame.fileName && frame.fileName.includes("/node_modules/webdriverio/"));
+    const isIgnoredFunction = (frame: StackFrame): boolean =>
+        frame.functionName
+            ? Boolean(ignoredFunctions.find(fn => fn === frame.functionName || "async " + fn === frame.functionName))
+            : false;
+    const shouldIncludeFrame = (frame: StackFrame): boolean => !isWdioFrame(frame) || !isIgnoredFunction(frame);
+
+    const framesFiltered = rawFramesArr.filter((_, i) => shouldIncludeFrame(framesParsed[i])).join("\n");
+
+    return applyStackFrames(error, framesFiltered);
+}
+
+export class ShallowStackFrames {
+    private _framesMap: Map<string, string>;
+    private _key: number;
+
+    constructor() {
+        this._framesMap = new Map();
+        this._key = 1;
+    }
+
+    getKey(): string {
+        return String(this._key++);
+    }
+
+    enter(key: string, frames: string): void {
+        this._framesMap.set(key, frames);
+    }
+
+    leave(key: string): void {
+        this._framesMap.delete(key);
+    }
+
+    isNested(childFrames: RawStackFrames): boolean {
+        for (const parentFrames of this._framesMap.values()) {
+            if (childFrames.length !== parentFrames.length && childFrames.endsWith(parentFrames)) {
```

Comment:
> childFrames.length !== parentFrames.length

Because user might have a loop.

> childFrames.endsWith(parentFrames)

If true, then we already wrapped parent call